src/hotspot/share/runtime/handshake.cpp

```diff
@@ -349,7 +349,16 @@ void HandshakeOperation::do_handshake(JavaThread* thread) {
 }
 
 void Handshake::execute(HandshakeClosure* hs_cl) {
-  HandshakeOperation cto(hs_cl, nullptr, Thread::current());
+  Thread* current_thread = Thread::current();
+
+  // The current thread must not belong to the SuspendibleThreadSet, because an
+  // on-the-fly safepoint can be waiting for the current thread, and the
+  // current thread will be blocked in VMThread::execute, resulting in
+  // deadlock.
+  assert(!current_thread->is_suspendible_thread(), "precondition");
+  assert(!current_thread->is_indirectly_suspendible_thread(), "precondition");
```

Comment:
Why not move the asserts to VMThread::execute() to cover any operation, not just VM_HandshakeAllThreads?

Reply:
Aren't Handshakes a superset of VMoperations? `VMThread::execute` will only affect VM_Operations, not direct Handshakes with other threads. Is the problem limited to a STS thread trying to execute a safepoint VM operation?

Reply:
There are two types of handshakes, the ones that target a single JavaThread and the ones that are executed on all JavaThreads. The latter one is a VM operation (VM_HandshakeAllThreads), i.e it’s executed by the VMThread. The original fix guarded against the deadlock scenario where the STS thread requested a VM_HandshakeAllThreads operation, but the same issue can happen by requesting any VM operation. As for direct handshakes, the original fix didn’t touch that code path, and I don’t think we need to because the handshaker can only be a JavaThread already, and AFAIK all these STS threads are not.

Reply:
Thanks Patricio. I had things inverted.